app/models/automation_step.rb

```diff
@@ -0,0 +1,90 @@
+# frozen_string_literal: true
+
+class AutomationStep < ApplicationRecord
+  belongs_to :automation
+  belongs_to :pipeline
+  belongs_to :launched_by, class_name: 'User', optional: true
+  has_one :pipeline_job, dependent: :destroy
+
+  validates :position, presence: true, numericality: { only_integer: true, greater_than_or_equal_to: 0 }
+
+  serialize :harvest_definition_ids, type: Array
+
+  def harvest_definitions
+    HarvestDefinition.where(id: harvest_definition_ids)
+  end
+
+  def display_name
+    "#{position + 1}. #{pipeline.name}"
+  end
+
+  # Returns the status of this step
+  # Prioritizes statuses in a logical order
+  def status
+    return 'not_started' if no_reports?
+
+    statuses = report_statuses
+
+    status_from_statuses(statuses)
+  end
+
+  def next_step
+    automation.automation_steps.where('position > ?', position).order(position: :asc).first
+  end
+
+  # Get destination from the automation
+  delegate :destination, to: :automation
+
+  # For compatibility with existing code that might expect destination_id
+  def destination_id
+    destination&.id
+  end
+
+  private
+
+  def no_reports?
+    pipeline_job&.harvest_reports.blank?
+  end
+
+  def report_statuses
+    return [] unless pipeline_job
+
+    reports = pipeline_job.harvest_reports
+    reports&.map(&:status)&.uniq || []
+  end
+
+  def status_from_statuses(statuses)
```

Comment:
This method and the status method in the class above are the same. I wonder if we can put the shared status logic somewhere?

Reply:
nice! all done, moved to a concern :D